dvc/path_info.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import unicode_literals
+
+import sys
+import os
+import posixpath
+from funcy import cached_property
+
+from dvc.utils.compat import str, basestring, is_py2
+from dvc.utils.compat import pathlib, fspath_py35, urlparse
+
+
+if is_py2:
+    fs_encoding = sys.getfilesystemencoding()
+
+
+class PathInfo(pathlib.PurePath):
+    scheme = "local"
+
+    def __new__(cls, *args):
+        # Construct a proper subclass depending on current os
+        if cls is PathInfo:
+            cls = WindowsPathInfo if os.name == "nt" else PosixPathInfo
+        return cls._from_parts(args)
+
+    @classmethod
+    def from_posix(cls, s):
+        return cls(pathlib.PurePosixPath(s))
+
+    def __str__(self):
+        path = self.__fspath__()
+        if os.name == "nt" and not os.path.commonprefix([os.getcwd(), path]):
+            # In case of windows, when cache is on different drive than
+            # workspace, we will get ValueError when trying to get relpath
+            return path
+        return os.path.relpath(path)
+
+    def __repr__(self):
+        return "{}: '{}'".format(type(self).__name__, self)
+
+    # This permits passing it to file utils directly in Python 3.6+
+    # With Python 2.7, Python 3.5+ we are stuck with path_info.fspath for now
+    def __fspath__(self):
+        return pathlib.PurePath.__str__(self)
+
+    @property
+    def fspath(self):
+        return self.__fspath__()
+
+    url = fspath
+
+    def relpath(self, other):
+        return self.__class__(
```

Comment:
don't we hit some Windows issues here as well?

Reply:
It would be callers issue.

Reply:
@Suor Well we are the "callers" :) Take a look at https://github.com/iterative/dvc/blob/master/dvc/path/local.py#L11 . I don't see the same thing in your patch.

Reply:
@Suor sorry, didn't notice in in __str__